# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/EnvironmentObject.cpp
# Commit: ea1bb2f0fd39
# Full Hash: ea1bb2f0fd39a412480bd777ba944b276a54006a
# Author: Logan Smyth <loganfsmyth@gmail.com>
# Date: 2019-12-09 09:50:39
# Regressor Bug: 1600204
# File Overlap Count: 1
# Description:
#   Bug 1600204 - Part 5: Support .environment on suspended generator frames. r=jimb
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D55235
# ==============================================================================

diff -r 39d1ae1352b3 -r ea1bb2f0fd39 js/src/vm/EnvironmentObject.cpp
--- a/js/src/vm/EnvironmentObject.cpp	Fri Dec 06 02:30:30 2019 +0000
+++ b/js/src/vm/EnvironmentObject.cpp	Sun Dec 08 23:53:27 2019 +0000
@@ -3261,6 +3261,16 @@
   return GetDebugEnvironment(cx, ei);
 }
 
+JSObject* js::GetDebugEnvironmentForSuspendedGenerator(
+    JSContext* cx, JSScript* script, AbstractGeneratorObject& genObj) {
+  RootedObject env(cx);
+  RootedScope scope(cx);
+  GetSuspendedGeneratorEnvironmentAndScope(genObj, script, &env, &scope);
+
+  EnvironmentIter ei(cx, env, scope);
+  return GetDebugEnvironment(cx, ei);
+}
+
 JSObject* js::GetDebugEnvironmentForFrame(JSContext* cx, AbstractFramePtr frame,
                                           jsbytecode* pc) {
   cx->check(frame);
@@ -3769,6 +3779,16 @@
   return true;
 }
 
+void js::GetSuspendedGeneratorEnvironmentAndScope(
+    AbstractGeneratorObject& genObj, JSScript* script, MutableHandleObject env,
+    MutableHandleScope scope) {
+  env.set(&genObj.environmentChain());
+
+  jsbytecode* pc =
+      script->offsetToPC(script->resumeOffsets()[genObj.resumeIndex()]);
+  scope.set(script->innermostScope(pc));
+}
+
 #ifdef DEBUG
 
 typedef HashSet<PropertyName*> PropertyNameSet;